### ISSUE TITLE
4.3.1 introduce a new bug: missing dep (tape)

### ISSUE THREAD

**Issue by MoOx** [stylelint/stylelint#738]
https://travis-ci.org/MoOx/statinamic/jobs/109072551#L1038

**Comment by MoOx** [comment]
@jeddy3 this time I am not sure how to fix this :/

**Comment by jeddy3** [comment]
Move tape to deps for now for hotfix, and then decide if other resolution is needed later?

**Comment by MoOx** [comment]
Yeah? Not really sure why tape is needed but I guess that's a solution since your added ruleTester as a public API.

**Comment by davidtheclark** [comment]
Yes. If we're going to expose the ruleTester (which I think we should), tape will need to be a dependency. I missed this.

Another possible approach is to figure out how to publish different npm packages from the one repo, and publish ruleTester as its own package so stylelint doesn't need the extra dependency.

**Comment by MoOx** [comment]
Or rely on native "assert" for this part (just an idea)?

**Comment by davidtheclark** [comment]
I am not super thrilled about the implementation of the rule tester, actually, but an scared that any attempt to change it might mean a huge amount of drudgery it since there are almost 10000 unit tests ༼ ºººººل͟ººººº ༽ .... so I don't want to spend hours converting to a new better system unless I'm really sure the new system is great.

Here are some problems:

- Want to ensure that it is absolutely as easy as possible to write tests, since there have to be so many.
- Want to ensure that when a test fails it is as easy as possible to find and understand the failure.
- Want to share the testing system so that plugins can use it.

One idea I had recently was to change the signature of the rule testing. Instead the ruleTester being a function (imitating tape) that runs a callback that it provides with an assertion object (`tr.ok()` and `tr.notOk()`), we could just create a schema for test descriptions. Then the `ruleTester` would be a function that simply accepts the schame. Something like:

```js
testRule({
  ruleName: "block-no-empty",
  config: true,
  pass: [ // array of objects describing tests that DO NOT trigger stylelint warnings
    { input: "" },
    { input: "@import \"foo.css\";" },
    { input: "a { color: pink; }" },
    { input: "@media print { a { color: pink; } }" },
    { input: "@import url(x.css)" },
  ],
  fail: [ // array of objects describing tests that DO trigger stylelint warnings
    {
      input: "a {}",
      message: messages.rejected,
      line: 1,
      column: 3,
    },
    {
      input: "a { }",
      message: messages.rejected,
      line: 1,
      column: 3,
    },
    {
      input: "a {\n}",
      message: messages.rejected,
      line: 1,
      column: 3,
    },
    {
      input: "@media print {}",
      message: messages.rejected,
      line: 1,
      column: 14,
    },
    {
      input: "@media print { a {} }",
      message: messages.rejected,
      line: 1,
      column: 18,
    },
  ],
})
```

What I think might be the big advantage of this approach is that the implementation is then totally open. You could write a tester in tape, another in ava, another in mocha, etc.: all it has to do it understand and properly test the schema. I was thinking about this because I wondered if the tests would run faster if I used ava --- but it was a huge pain to try.

Or we could write it only as an assertion engine, and then wrap it in another testing library for our tests. For example,

```js
import test from "tape"
test("block-no-empty", () => {
  stylelintAssert({ /* schema here */ })
  stylelintAssert({ /* schema here */ })
  /* ... */
})
```

Ooo, I kind of like that idea. What do you think?

**Comment by davidtheclark** [comment]
Made a proof of concept. Here's what a test could look like (using Mocha):

```js
import createStylelintAssert from "../../../testUtils/stylelintAssert"

import rule, { ruleName, messages } from ".."

const stylelintAssert = createStylelintAssert()

describe(ruleName, () => {
  it("true", done => {
    stylelintAssert(rule, {
      ruleName,
      config: true,
      accept: [
        { code: "" },
        { code: "@import \"foo.css\";" },
        { code: "a { color: pink; }" },
        { code: "@media print { a { color: pink; } }" },
        { code: "@import url(x.css)" },
      ],
      reject: [
        {
          code: "a {}",
          message: messages.rejected,
          line: 2,
          column: 3,
        },
        {
          code: "a { }",
          message: messages.rejected,
          line: 1,
          column: 3,
        },
        {
          code: "a {\n}",
          message: messages.rejected,
          line: 1,
          column: 3,
        },
        {
          code: "@media print {}",
          message: messages.rejected,
          line: 1,
          column: 14,
        },
        {
          code: "@media print { a {} }",
          message: messages.rejected,
          line: 1,
          column: 18,
        },
      ],
    }, done)
  })
})
```

And here's the output (notice there's a failing test in there):

```

  block-no-empty
    1) true


  0 passing (69ms)
  1 failing

  1) block-no-empty true:

      AssertionError: 

> rule: block-no-empty
> config: true
> code: "a {}"
> should register warning with line 2
> warnings:
- 1.3: Unexpected empty block (block-no-empty)

      + expected - actual

      -1
      +2
      
      at stylelintAssert.js:65:15
```

Why using mocha? tape makes kind of unreadable output (does not allow for new lines unless you do something very tape-specific), but could still work with this interface. With tape I got this:

```
TAP version 13
# block-no-empty
ok 1  > rule: block-no-empty > config: true > code: "" > should be accepted > warnings: 
ok 2  > rule: block-no-empty > config: true > code: "@import \"foo.css\";" > should be accepted > warnings: 
ok 3  > rule: block-no-empty > config: true > code: "a { color: pink; }" > should be accepted > warnings: 
ok 4  > rule: block-no-empty > config: true > code: "@media print { a { color: pink; } }" > should be accepted > warnings: 
ok 5  > rule: block-no-empty > config: true > code: "@import url(x.css)" > should be accepted > warnings: 
ok 6  > rule: block-no-empty > config: true > code: "a {}" > should be rejected with one warning > warnings: - 1.3: Unexpected empty block (block-no-empty) 
ok 7  > rule: block-no-empty > config: true > code: "a {}" > should register warning with message "Unexpected empty block (block-no-empty) > warnings: - 1.3: Unexpected empty block (block-no-empty) 
not ok 8  > rule: block-no-empty > config: true > code: "a {}" > should register warning with line 2 > warnings: - 1.3: Unexpected empty block (block-no-empty) 
  ---
    operator: equal
    expected: 2
    actual:   1
    at: run (/usr/local/lib/node_modules/babel-tape-runner/node_modules/core-js/modules/es6.promise.js:104:47)
  ...
ok 9  > rule: block-no-empty > config: true > code: "a {}" > should register warning with column 3 > warnings: - 1.3: Unexpected empty block (block-no-empty) 
ok 10  > rule: block-no-empty > config: true > code: "a { }" > should be rejected with one warning > warnings: - 1.3: Unexpected empty block (block-no-empty) 
ok 11  > rule: block-no-empty > config: true > code: "a { }" > should register warning with message "Unexpected empty block (block-no-empty) > warnings: - 1.3: Unexpected empty block (block-no-empty) 
ok 12  > rule: block-no-empty > config: true > code: "a { }" > should register warning with line 1 > warnings: - 1.3: Unexpected empty block (block-no-empty) 
ok 13  > rule: block-no-empty > config: true > code: "a { }" > should register warning with column 3 > warnings: - 1.3: Unexpected empty block (block-no-empty) 
ok 14  > rule: block-no-empty > config: true > code: "a {\n}" > should be rejected with one warning > warnings: - 1.3: Unexpected empty block (block-no-empty) 
ok 15  > rule: block-no-empty > config: true > code: "a {\n}" > should register warning with message "Unexpected empty block (block-no-empty) > warnings: - 1.3: Unexpected empty block (block-no-empty) 
ok 16  > rule: block-no-empty > config: true > code: "a {\n}" > should register warning with line 1 > warnings: - 1.3: Unexpected empty block (block-no-empty) 
ok 17  > rule: block-no-empty > config: true > code: "a {\n}" > should register warning with column 3 > warnings: - 1.3: Unexpected empty block (block-no-empty) 
ok 18  > rule: block-no-empty > config: true > code: "@media print {}" > should be rejected with one warning > warnings: - 1.14: Unexpected empty block (block-no-empty) 
ok 19  > rule: block-no-empty > config: true > code: "@media print {}" > should register warning with message "Unexpected empty block (block-no-empty) > warnings: - 1.14: Unexpected empty block (block-no-empty) 
ok 20  > rule: block-no-empty > config: true > code: "@media print {}" > should register warning with line 1 > warnings: - 1.14: Unexpected empty block (block-no-empty) 
ok 21  > rule: block-no-empty > config: true > code: "@media print {}" > should register warning with column 14 > warnings: - 1.14: Unexpected empty block (block-no-empty) 
ok 22  > rule: block-no-empty > config: true > code: "@media print { a {} }" > should be rejected with one warning > warnings: - 1.18: Unexpected empty block (block-no-empty) 
ok 23  > rule: block-no-empty > config: true > code: "@media print { a {} }" > should register warning with message "Unexpected empty block (block-no-empty) > warnings: - 1.18: Unexpected empty block (block-no-empty) 
ok 24  > rule: block-no-empty > config: true > code: "@media print { a {} }" > should register warning with line 1 > warnings: - 1.18: Unexpected empty block (block-no-empty) 
ok 25  > rule: block-no-empty > config: true > code: "@media print { a {} }" > should register warning with column 18 > warnings: - 1.18: Unexpected empty block (block-no-empty) 

1..25
# tests 25
# pass  24
# fail  1
```

The idea here is that `stylelintAssert()` would just make assertions and allow you to use the test runner of your choice. Actually you can feed in the equality assertion, too --- that's how the tape example works.

**Comment by jeddy3** [comment]
I like this. I'm probably the least qualified to talk about testing though...

Is it possible to see what the output is like if the runner of choice (used internally within stylelint) is `ava`?

**Comment by davidtheclark** [comment]
ava has some bug currently that is causing the output to look crappy if you use native asserts. But looks like it might be fixed in next release.

**Comment by davidtheclark** [comment]
Well. This issue is "resolved" in 4.3.3, when the tape dependency was dropped. I'm going to close this issue, and then open something new if I figure out a new and improved method for rule testing and sharing the rule tester.
